integration/read_write_mode_test.go

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+//go:build requires_docker
+// +build requires_docker
+
+package integration
+
+import (
+	"math"
+	"testing"
+	"time"
+
+	"github.com/grafana/e2e"
+	e2edb "github.com/grafana/e2e/db"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/stretchr/testify/require"
+
+	"github.com/grafana/mimir/integration/e2emimir"
+)
+
+func TestReadWriteMode(t *testing.T) {
+	s, err := e2e.NewScenario(networkName)
+	require.NoError(t, err)
+	defer s.Close()
+
+	minio := e2edb.NewMinio(9000, blocksBucketName, alertsBucketName, rulerBucketName)
+	require.NoError(t, s.StartAndWaitReady(minio))
+
+	flags := map[string]string{
+		"-common.storage.backend":              "s3",
+		"-common.storage.s3.access-key-id":     e2edb.MinioAccessKey,
+		"-common.storage.s3.secret-access-key": e2edb.MinioSecretKey,
+		"-common.storage.s3.endpoint":          minio.NetworkHTTPEndpoint(),
+		"-common.storage.s3.insecure":          "true",
+		"-blocks-storage.s3.bucket-name":       blocksBucketName,
+		"-alertmanager-storage.s3.bucket-name": alertsBucketName,
+		"-ruler-storage.s3.bucket-name":        rulerBucketName,
+		"-ingester.ring.replication-factor":    "1",
+		"-memberlist.join":                     "mimir-backend-1",
+	}
+
+	readInstance := e2emimir.NewReadInstance("mimir-read-1", flags)
+	writeInstance := e2emimir.NewWriteInstance("mimir-write-1", flags)
+	backendInstance := e2emimir.NewBackendInstance("mimir-backend-1", flags)
+	require.NoError(t, s.StartAndWaitReady(readInstance, writeInstance, backendInstance))
+
+	c, err := e2emimir.NewClient(writeInstance.HTTPEndpoint(), readInstance.HTTPEndpoint(), "", "", "user-1")
+	require.NoError(t, err)
+
+	// Push some data to the cluster.
+	now := time.Now()
+	series, expectedVector := generateSeries("test_series_1", now, prompb.Label{Name: "foo", Value: "bar"})
+
+	res, err := c.Push(series)
+	require.NoError(t, err)
+	require.Equal(t, 200, res.StatusCode)
+
+	// Verify we can read the data we just pushed, both with an instant query and a range query.
+	result, err := c.Query("test_series_1", now)
+	require.NoError(t, err)
+	require.Equal(t, model.ValVector, result.Type())
+	require.Equal(t, expectedVector, result.(model.Vector))
+
+	_, err = c.QueryRange("test_series_1", now.Add(-5*time.Minute), now, 15*time.Second)
+	require.NoError(t, err)
```

Comment:
Other integration tests seem split 50:50 between just checking that a range query doesn't return an error and actually checking the response is as expected - is there value in checking that a range query returns the expected data in this scenario? (why / why not?)

Reply:
Asserting on query results is always better. Some tests don't do it because it's annoying to generate the expected value matrix (even if now we have an utility like `generateExpectedValues()`). Given your great work on this PR, willing to take a look at all integration tests calling `QueryRange()` and asserting on the results, eventually building utilities to conveniently generate the expected result? If in some cases generate stable expected results is difficult, you can just skip asserting on the results in such cases.

Reply:
Yep sure, happy to try to tidy this up everywhere - there are a couple of things like this that I would like to improve in all of the tests (others include your comment about the magic time values for the label names and values queries, and replacing the string interpolation to construct the address for Minio with using `minio.NetworkHTTPEndpoint()`)

Reply:
Take a look at #3433 (and particularly https://github.com/grafana/mimir/pull/3433/files#r1019837210) and let me know what you think.